### ISSUE TITLE
Pilot, Xcode 8 and Automatic signing. ERROR ITMS-90161

### ISSUE THREAD

**Issue by poolqf** [fastlane/fastlane#6191]
### New Issue Checklist
- [x] Updated fastlane to the latest version
- [x] I have read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
### Issue Description

It has been impossible to upload a build with Pilot, using Xcode 8 and automatic signing. It continuously says `ERROR ITMS-90161: "Invalid Provisioning Profile. The provisioning profile included in the bundle com.***.*** [Payload/***.app] is invalid.`

if I try using `deliver`, the same error message appears.

Things checked:
- When `gym` creates the build, it has the correct bundle id
- When `gym` creates the build, it appears in the Xcode 8 organizer. **If I upload through there, everything works**
##### Complete output when running fastlane, including the stack trace and command used

```
[00:42:23]: Successfully exported and compressed dSYM file
[00:42:23]: Successfully exported and signed the ipa file:
[00:42:23]: /Users/polquintana/Xcode/***/***.ipa
[00:42:23]: -----------------------------------------------
[00:42:23]: --- Step: Switch to ios checkout_badge lane ---
[00:42:23]: -----------------------------------------------
[00:42:23]: Cruising over to lane 'ios checkout_badge' 🚖
[00:42:23]: Cruising back to lane 'ios beta' 🚘
[00:42:23]: -------------------
[00:42:23]: --- Step: pilot ---
[00:42:23]: -------------------
[00:42:23]: Login to iTunes Connect (***)
[00:42:32]: Login successful
[00:42:32]: Ready to upload new build to TestFlight (App: ***)...
[00:42:33]: Going to upload updated app to iTunes Connect
[00:42:33]: This might take a few minutes. Please don't interrupt the script.
[00:44:47]: [Transporter Error Output]: ERROR ITMS-90161: "Invalid Provisioning Profile. The provisioning profile included in the bundle com.***.*** [Payload/***.app] is invalid. [Missing code-signing certificate]. A Distribution Provisioning profile should be used when submitting apps to the App Store. For more information, visit the iOS Developer Portal."
[00:44:47]: Transporter transfer failed.
[00:44:47]:
[00:44:47]: ERROR ITMS-90161: "Invalid Provisioning Profile. The provisioning profile included in the bundle com.***.*** [Payload/***.app] is invalid. [Missing code-signing certificate]. A Distribution Provisioning profile should be used when submitting apps to the App Store. For more information, visit the iOS Developer Portal."
[00:44:47]: ERROR ITMS-90161: "Invalid Provisioning Profile. The provisioning profile included in the bundle com com.***.*** [Payload/***.app] is invalid. [Missing code-signing certificate]. A Distribution Provisioning profile should be used when submitting apps to the App Store. For more information, visit the iOS Developer Portal."
Return status of iTunes Transporter was 1: ERROR ITMS-90161: "Invalid Provisioning Profile. The provisioning profile included in the bundle com.***.*** [Payload/***.app] is invalid. [Missing code-signing certificate]. A Distribution Provisioning profile should be used when submitting apps to the App Store. For more information, visit the iOS Developer Portal."
The call to the iTMSTransporter completed with a non-zero exit status: 1. This indicates a failure.
[00:44:47]: -------------------
[00:44:47]: --- Step: slack ---
[00:44:47]: -------------------
[00:44:50]: Successfully sent Slack notification
[00:44:50]: Variable Dump:
[00:44:50]: {:DEFAULT_PLATFORM=>:ios, :PLATFORM_NAME=>:ios, :LANE_NAME=>"ios beta", :IPA_OUTPUT_PATH=>"/Users/polquintana/Xcode/***/***.ipa", :DSYM_OUTPUT_PATH=>"/Users/polquintana/Xcode/***/***.app.dSYM.zip", :XCODEBUILD_ARCHIVE=>"/Users/polquintana/Library/Developer/Xcode/Archives/2016-09-20/*** 2016-09-20 00.40.57.xcarchive"}
[00:44:50]: Error uploading ipa file, for more information see above

+------+-------------------------------------+-------------+
|                     fastlane summary                     |
+------+-------------------------------------+-------------+
| Step | Action                              | Time (in s) |
+------+-------------------------------------+-------------+
| 1    | Verifying required fastlane version | 0           |
| 2    | default_platform                    | 0           |
| 3    | Switch to ios custom_badge lane     | 0           |
| 4    | badge                               | 5           |
| 5    | gym                                 | 87          |
| 6    | Switch to ios checkout_badge lane   | 0           |
| 7    | pilot                               | 144         |
| 8    | slack                               | 2           |
+------+-------------------------------------+-------------+

[00:44:50]: fastlane finished with errors

[!] Error uploading ipa file, for more information see above
```
### Configuration Files

Please copy the complete content of your `Fastfile` and any other configuration files you use below:

Fastfile:

``` ruby
  desc "Submit a new Beta Build to Apple TestFlight"
  lane :beta do
    ensure_git_branch
    custom_badge
    gym(scheme: "***", configuration: "Beta", use_legacy_build_api: true)
    checkout_badge
    pilot(app_identifier: "com.***.***", skip_submission: true, skip_waiting_for_build_processing: true)
    slack(message: "Deployment to TestFlight finished.")
  end
```
### Environment

fastlane version (run `fastlane -v`): fastlane 1.103.0

Do you use bundler to execute fastlane (i.e. `bundle exec fastlane`)? no

Do you use a Ruby environment manager (e.g. `chruby`, `rbenv`, `rvm`)? rvm

**Comment by KrauseFx** [comment]
Please follow the code signing guide posted by the bot, we recommend disabling automatic code signing and letting fastlane manage the certificates for you

**Comment by poolqf** [comment]
Moving to `match` fixed the issue, but would be interesting to have support for automatic signing. Is this on the roadmap?

**Comment by TKBurner** [comment]
@poolqf Thank for following up. This isn't on the roadmap with the most recent changes to Xcode. Thanks!

**Comment by KrauseFx** [comment]
@poolqf If you use automatic code signing which is provided by Xcode, you don't need _match_. How would you like the two to work together?

**Comment by poolqf** [comment]
@KrauseFx the ideal solution for me would be to let Xcode do all the work, and not use match. 

I feel like match is great when you are in an app that several people is developing. I don't have experience using Xcode's automatic signing on an app with multiple developers, maybe it also does all the work for you.

**Comment by KrauseFx** [comment]
@poolqf _match_ will sync your private keys also, which Xcode doesn't. So it doesn't cover distribution

**Comment by ossmalpha** [comment]
Still nothing works. Can't upload using fastlane, but works using Organizer in Xcode. :(

**Comment by TKBurner** [comment]
@ossmalpha Are you also affected by this issue? Can you provide your full input and output? Thanks!

**Comment by mangofever** [comment]
I'm experiencing this issue. I don't use match and just set automatically manage signing. And it works uploading ipa file via organazer.

**Comment by TKBurner** [comment]
@mangofever Thanks for reaching out. We don't currently support automatic signing and this could be the source of your problem. Also, can you let me know the full input and output of your error? Thanks!

**Comment by mangofever** [comment]
@TKBurner. 
As you said, it was because of automatic signing.
After I changed back to manual signing, the error just disappeared.
We are currently using 2 different Apple accounts. (One is for App Store, and another for Enterprise)

Input is just a username of App Store account like...
pilot(
     username: APP_STORE_ACCOUNT
)

[Error output]
[17:05:15]: -------------------
[17:05:15]: --- Step: pilot ---
[17:05:15]: -------------------
[17:05:15]: Login to iTunes Connect (xxx@xxxxxx.com)
[17:05:17]: Login successful
[17:05:17]: Ready to upload new build to TestFlight (App: xxxxxxxxxx)...
[17:05:20]: Going to upload updated app to iTunes Connect
[17:05:20]: This might take a few minutes. Please don't interrupt the script.
[17:07:22]: [Transporter Error Output]: ERROR ITMS-90161: "Invalid Provisioning Profile. The provisioning profile included in the bundle xxx.yyyyy.zzzzzz [Payload/xxxxxx.app] is invalid. [Missing code-signing certificate]. A Distribution Provisioning profile should be used when submitting apps to the App Store. For more information, visit the iOS Developer Portal."

[17:07:22]: Transporter transfer failed.
[17:07:22]: 
[17:07:22]: ERROR ITMS-90161: "Invalid Provisioning Profile. The provisioning profile included in the bundle xxx.yyyyy.zzzzzz [Payload/xxxxxx.app] is invalid. [Missing code-signing certificate]. A Distribution Provisioning profile should be used when submitting apps to the App Store. For more information, visit the iOS Developer Portal."

[17:07:22]: ERROR ITMS-90161: "Invalid Provisioning Profile. The provisioning profile included in the bundle xxx.yyyyy.zzzzzz [Payload/xxxxxx.app] is invalid. [Missing code-signing certificate]. A Distribution Provisioning profile should be used when submitting apps to the App Store. For more information, visit the iOS Developer Portal."
Return status of iTunes Transporter was 1: ERROR ITMS-90161: "Invalid Provisioning Profile. The provisioning profile included in the bundle xxx.yyyyy.zzzzzz [Payload/xxxxxx.app] is invalid. [Missing code-signing certificate]. A Distribution Provisioning profile should be used when submitting apps to the App Store. For more information, visit the iOS Developer Portal."
The call to the iTMSTransporter completed with a non-zero exit status: 1. This indicates a failure.

Hope this help.

**Comment by TKBurner** [comment]
@mangofever Really glad to hear you got up and running. We are looking into improving the way this fails. Thanks!

**Comment by vutiet** [comment]
Hi,
I'm facing this issue, too. My project has "Automatically manage signing" enabled and when it ran _pilot_ I has exact error as **mangofever** posted above. 
May I ask that you guys will have update soon? Or I have to go with manual signing? Thanks

**Comment by TKBurner** [comment]
@vutiet You should go with manual signing to have the most control. Here are our docs on automatic signing: https://docs.fastlane.tools/codesigning/xcode-project/#using-match-development-or-xcode-automatic-provisioning

Thanks!

**Comment by nicolasbraun** [comment]
Personally my projects settings are in Automatic by default for developers and I change to manual using simple regexp on the pbxproj (for delivery jobs by Jenkins using fastlane)

**Comment by CurlyHeir** [comment]
@nicolasbraun 
could you write about this solution somewhere?

**Comment by nicolasbraun** [comment]
Sure !
What we do: 

* Enable automatic codesign for the target
* Use our bundle for all configuration but Distribution https://i.imgur.com/SNrRveD.png

Automatic codesign is commited for development so the developers do not have to worry about it, only the PMs for developer or the CI servers.

At the beggining of the Fastlane flow, we:

**Switch to Manual codesign**

To do that I have a simple function 

```ruby
def switch_to_manual_codesign
  puts "Switching to manual codesign".yellow
  pbxproj = '../' + ENV['XCODEPROJ'] + '/project.pbxproj'
  text = File.read(pbxproj)
  new_contents = text.gsub( /(ProvisioningStyle = )Automatic(;)/, '\1' + "Manual" +'\2')
  File.open(pbxproj, "w") {|file| file.puts new_contents }
end
```
*NB : I suspect we could only replace Automatic with manual but this seems safer*

**Switch to Production Signing (if required)**

By default Automatic codesign uses Development, so if my lanes requires it (totally depends on your setup)

```ruby
def switch_to_production_codesign
  puts "Switching to production codesign".yellow
  pbxproj = '../' + ENV['XCODEPROJ'] + '/project.pbxproj'
  text = File.read(pbxproj)
  new_contents = text.gsub('iPhone Developer', 'iPhone Distribution')
  File.open(pbxproj, "w") {|file| file.puts new_contents }
end
```

I did not take time to factorize this yet :)

Feel free to ask!

Nicolas

**Comment by TKBurner** [comment]
@nicolasbraun Thanks so much for sharing!

**Comment by vutiet** [comment]
Thanks guys, a lot of insights but unfortunately I couldn't make it with manual signing because my project has extension app (i.e. watch app). So, I had a workaround as using automatic signing to export .ipa file then using _resign_ to sign it with production provisioning profile before upload it to testflight.

**Comment by nicolasbraun** [comment]
Hi @vutiet 

I also have some projects with extensions. Can you elaborate on what is blocking you?

**Comment by vutiet** [comment]
Hi @nicolasbraun, 
my project settings are as below: 
- iPhone app has bundle id com.companyname.iphoneapp with provisioning profile AppStore_com.companyname.iphoneapp
- watchkit extension has bundle id com.companyname.watchkitextension with provisioning profile AppStore_com.companyname.watchkitextension
- watch app has bundle id com.companyname.watchapp with provisioning profile AppStore_com.companyname.watchapp

So if I switch to manual signing, I can set the provisioning profile for only iPhone app, watchkitextension and watchapp haven't got set and gym will fail

**Comment by nicolasbraun** [comment]
What we do is use eVar in xcode:

$PROFILE_UDID for the app https://i.imgur.com/huI88lS.png 
$PROFILE_UDID_EXT0 / EXT1 for the watch app and extension. https://i.imgur.com/UGhNU7l.png

We then run match 3 times and set the correct eVar fort each and 🚀 .

That said we have to do this because we have several bundle for internal development and for the release on our client's account. I believe since Xcode8 can now use Profile name instead of UUID you could simple set the value to the profile name (which does not change when you update it)

Nicolas

**Comment by sahandnayebaziz** [comment]
I'm also affected by this issue and would love to leave signing configuration to Xcode and the automation of versioning through to pushing to iTunes Connect to fastlane. As a solo developer, I'm not personally attracted to `match` but see the value it offers teams.

Cert/Sigh have not worked as consistently for me as everything did in Xcode 7 without cert/sigh, so for now I will be returning to manually archiving/submitting in Xcode. Hopefully soon this becomes stable again without having to move to match.

Thank you for great software nonetheless

**Comment by pahnev** [comment]
We have also been having some issues letting Xcode manage our profiles. We started at the end of 2016 and it seemed to be working fine, with different schemes and multiple extensions and 2 devs.

However at some point the other dev was able to only make 1/2 builds and so was I, but they were the opposite ones, so we managed. Then I left for vacations and both builds started working for him and when I came back neither of the schemes worked for me anymore.

Now we have added a new dev to the project and us "old devs" are not able to create any builds using fastlane because the gym is failing with error _"No provisioning profiles with a valid signing identity"_ for all the extensions we have. So I took a bit deeper look into this.

In the dev center I noticed that all the distribution profiles were "invalid".
![screen shot 2017-03-09 at 14 59 18](https://cloud.githubusercontent.com/assets/2078494/23751741/48f44870-04db-11e7-9d31-a8d0983ca922.png)

I tried to re-generate them(or whatever happens when you edit the profile and just press generate) to make them active again and downloading those valid profiles. With the downloaded profiles fastlane failed with error message _"Code signing is required for product type 'App Extension' in SDK 'iOS 10.2'"._
If I opened Xcode, it would start making "repairs" and those profiles would get invalidated again, so for whatever reason those new "active" profiles were no good.

I then tried using `sigh` and it actually fixed one of the schemes for me, but the other one still errored out because of the app extension provisioning profiles. So logically I added sigh for all the app extensions and it actually started working!

So to me it seems, that while the Xcode's automatic management is nice if you use Xcode to create builds, it totally breaks fastlane, by being a bit too magical.
